From 1e994acd2c1fb0194e0b5e5a2e00454e0c0575e7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Arkadiusz=20Mi=C5=9Bkiewicz?= <arekm@maven.pl>
Date: Thu, 22 Oct 2026 10:52:01 +0200
Subject: [PATCH] Fix #74779: x() and y() truncating floats to integers

With MYSQLI_OPT_INT_AND_FLOAT_NATIVE, mysqlnd converts text-protocol
floats with locale-dependent atof(), so under e.g. LC_NUMERIC=de_DE
"53.34" becomes 53.

Origin: upstream, https://github.com/php/php-src/commit/3f8d21b9226493365cbb9f913322f5da6965c974
Origin: upstream, https://github.com/php/php-src/commit/4ff2665509a0d2172e558fa21ed6718792349e85
Bug: https://bugs.php.net/74779
---
 ext/mysqli/tests/bug74779.phpt     | 42 ++++++++++++++++++++++++++++++
 ext/mysqlnd/mysqlnd_wireprotocol.c |  2 +-
 2 files changed, 43 insertions(+), 1 deletion(-)
 create mode 100644 ext/mysqli/tests/bug74779.phpt

diff --git a/ext/mysqli/tests/bug74779.phpt b/ext/mysqli/tests/bug74779.phpt
new file mode 100644
index 00000000..88e658ef
--- /dev/null
+++ b/ext/mysqli/tests/bug74779.phpt
@@ -0,0 +1,42 @@
+--TEST--
+Bug #74779 (x() and y() truncating floats to integers)
+--SKIPIF--
+<?php
+require_once('skipif.inc');
+require_once('skipifconnectfailure.inc');
+if (!setlocale(LC_NUMERIC, "de_DE", "de_DE.UTF-8", "de-DE")) die('skip locale not available');
+?>
+--FILE--
+<?php
+require_once("connect.inc");
+
+if (!$link = my_mysqli_connect($host, $user, $passwd, $db, $port, $socket)) {
+    printf("[001] Cannot connect to the server using host=%s, user=%s, passwd=***, dbname=%s, port=%s, socket=%s\n",
+        $host, $user, $db, $port, $socket);
+}
+
+if (!setlocale(LC_NUMERIC, "de_DE", "de_DE.UTF-8", "de-DE")) {
+    echo "[002] Cannot set locale\n";
+}
+
+if (!$link->options(MYSQLI_OPT_INT_AND_FLOAT_NATIVE, true)) {
+    printf("[003] [%d] %s\n", $link->errno, $link->error);
+}
+
+if (!$result = $link->query("SELECT ST_Y(Point(56.7, 53.34))")) {
+    printf("[004] [%d] %s\n", $link->errno, $link->error);
+}
+
+if (!$array = $result->fetch_array(MYSQLI_ASSOC)) {
+    printf("[005] [%d] %s\n", $link->errno, $link->error);
+}
+
+var_dump($array);
+
+mysqli_close($link);
+?>
+--EXPECT--
+array(1) {
+  ["ST_Y(Point(56.7, 53.34))"]=>
+  float(53,34)
+}
diff --git a/ext/mysqlnd/mysqlnd_wireprotocol.c b/ext/mysqlnd/mysqlnd_wireprotocol.c
index a405af0f..e9e342d4 100644
--- a/ext/mysqlnd/mysqlnd_wireprotocol.c
+++ b/ext/mysqlnd/mysqlnd_wireprotocol.c
@@ -1759,7 +1759,7 @@ php_mysqlnd_rowp_read_text_protocol_aux(MYSQLND_ROW_BUFFER * row_buffer, zval *
 				zend_uchar save = *(p + len);
 				/* We have to make it ASCIIZ temporarily */
 				*(p + len) = '\0';
-				ZVAL_DOUBLE(current_field, atof((char *) p));
+				ZVAL_DOUBLE(current_field, zend_strtod((char *) p, NULL));
 				*(p + len) = save;
 			} else
 #endif /* MYSQLND_STRING_TO_INT_CONVERSION */
-- 
2.56.0

